otwcode/otwarchive
# AO3-5801 Do not close self-closing tags for better HTML5 compliance

# Pull Request Checklist

  as the *first* thing in your pull request title (e.g. `AO3-1234 Fix thing`)
  until they are reviewed and merged before creating new pull requests.

## Issue

https://otwarchive.atlassian.net/browse/AO3-5801

## Purpose

- Some multimedia tags are self-closing in HTML5. So don't try to close them. ``<source ...>`` should remain ``<source ...>``, not ``<source ...></source>``, for example.
- This is an HTML4 vs HTML5 difference, and the current sanitizer uses HTML4 parser. So more broadly, move away from HTML4 and XHTML to HTML5. The HTML output has been declared to be HTML5 as a whole since AO3-4365 (#2217).
- To do so, prefer ``Nokogiri::HTML5`` over ``Nokogiri::HTML`` and ``Nokogiri::HTML4`` (but still use ``Nokogiri::HTML4`` where necessary). Note that ``Nokogiri::HTML`` and ``Nokogiri::HTML4`` are synonymous.
- The shift to HTML5 will bring other miscelleneous changes in how the sanitizer behaves. I hope to make the impact as small as possible and added changes that I believe are necessary, so this set of commits may be bigger than anticipated.(But I'll be honest, I might have been overzealous to make Rubocop happier in parts of the code that are only marginally related to this issue.)

## Testing Instructions

Trying to post and import works with audio/video tags would be a good place to start. However, I'd go beyond that.

This set of changes (at least its initial version) changes the inner working of the sanitization process at some fundamental level, so the scope of testing would probably have to be large. I tried to make the changes minimal and reasonable, but some unintended consequences might still happen. Known side effects are written above.

Posting and importing trigger sanitization, and I think that's mainly where testing is needed. It would be wise to test this as thoroughly as possible, using short and long, old and new content.

If possible, I would check performance metrics (like average time needed to complete one posting action) before and after this change. During my own (limited amount of) manual testing, there was no noticeable difference, though.

## References

- https://github.com/otwcode/otwarchive/pull/3678 by @sarken - I reused some of the test code.
- [AO3-5693](https://otwarchive.atlassian.net/browse/AO3-5693) (#3679) - a hack that should be made obsolete by this change
- [AO3-5974](https://otwarchive.atlassian.net/browse/AO3-5974) (#3847) - same as above

The Jira issue contains some preliminary discussion, too.
 
## Credit

Potpotkettle (they/them)
